executor/coprocessor.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/coprocessor"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/timeutil"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// CoprocessorDAGHandler uses to handle cop dag request.
+type CoprocessorDAGHandler struct {
+	sctx    sessionctx.Context
+	resp    *coprocessor.Response
+	selResp *tipb.SelectResponse
+	dagReq  *tipb.DAGRequest
+}
+
+// NewCoprocessorDAGHandler creates a new CoprocessorDAGHandler.
+func NewCoprocessorDAGHandler(sctx sessionctx.Context, resp *coprocessor.Response) *CoprocessorDAGHandler {
+	return &CoprocessorDAGHandler{
+		sctx:    sctx,
+		resp:    resp,
+		selResp: &tipb.SelectResponse{},
+	}
+}
+
+// HandleCopDAGRequest handles the coprocessor request.
+func (h *CoprocessorDAGHandler) HandleRequest(ctx context.Context, req *coprocessor.Request) *coprocessor.Response {
+	e, err := h.buildDAGExecutor(req)
+	if err != nil {
+		return h.buildResponse(err)
+	}
+
+	err = e.Open(ctx)
+	if err != nil {
+		return h.buildResponse(err)
+	}
+
+	chk := newFirstChunk(e)
+	tps := e.base().retFieldTypes
+	for {
+		chk.Reset()
+		err = Next(ctx, e, chk)
+		if err != nil {
+			break
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+		err = h.appendChunk(chk, tps)
+		if err != nil {
+			break
+		}
+	}
+	return h.buildResponse(err)
+}
+
+func (h *CoprocessorDAGHandler) buildDAGExecutor(req *coprocessor.Request) (Executor, error) {
+	if req.GetTp() != kv.ReqTypeDAG {
+		return nil, errors.Errorf("unsupported request type %d", req.GetTp())
+	}
+	dagReq := new(tipb.DAGRequest)
+	err := proto.Unmarshal(req.Data, dagReq)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	stmtCtx := h.sctx.GetSessionVars().StmtCtx
+	stmtCtx.SetFlagsFromPBFlag(dagReq.Flags)
+	stmtCtx.TimeZone, err = timeutil.ConstructTimeZone(dagReq.TimeZoneName, int(dagReq.TimeZoneOffset))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	h.dagReq = dagReq
+	return h.buildDAG(dagReq.Executors)
```

Comment:
Seems there is no need to extract an individual function, it’s better to inline function buildDAG to make logic more fluently.

Reply:
done.